BLE: Add ATT_MTU and Data Length docs

paul-szczepanek-arm opened the pull request.

paul-szczepanek-arm committed:
add BLE ATT_MTU and Data Length docs

pan- commented:
Corresponding PR in mbed-os: https://github.com/ARMmbed/mbed-os/pull/9537 .

pan- commented on docs/api/bluetooth/Gap.md:
```diff
@@ -52,6 +52,14 @@ You may set preferred PHYs (separately for RX and TX) using `setPreferredPhys()`
+
+In addition to modulation schemes, MTU (Maximum Transmission Unit) size also strongly affects throughput. Newer controllers will allow you to negotiate bigger MTUs. Since each packet contains overhead bigger packets maximise throughput.
+
+There are two separate MTUs to consider: the ATT_MTU (maximum attribute size) and Data Length. ATT_MTU is dealt with in `GattServer` and `GattClient`. `Gap` only deals with Data Length which is the maximum size of the packet that carries attributes which are fragmented across many such packets. ATT_MTU and Data Length are independent of each other.
```
Would it be possible to mention ATT_MTU at the end of the Data Length section rather than at the start of the second paragraph ?

pan- commented on docs/api/bluetooth/GattClient.md:
```diff
@@ -16,6 +16,14 @@ As a result of the discovery process, the client can start interacting with the
 
 Mbed BLE abstracts read and write operations to offer a single API that can be used to read or write characteristic values. The application code does not have to handle the necessary fragmentation or reassembly process if the attribute value to be transported can't fit in a single data packet.
 
+#### ATT_MTU
```
Maybe we can rework the section name into something less cryptic. What do you think of _GATT Maximum Transmission Unit (MTU)_ ?

paul-szczepanek-arm replied:
how about Attribute Maximum Transmission Unit (ATT_MTU)

pan- replied:
I'm not sure: we do not even define what GATT stands for in this file; `Attribute` seems a bit alien to me given the context.

paul-szczepanek-arm replied:
But it is technically correct. The packet is an Attribute Protocol packet. I'm taking the terms from the spec to be consistent with other literature.

pan- replied:
I do not deny that. I'm just pointing that it would be good to introduce the term Attribute Protocol (ATT) and Generic Attribute Profile (GATT) before the title. But we can go with _Attribute Protocol Maximum Transmission Unit_

paul-szczepanek-arm replied:
I introduced the acronyms and the explanation of their relationship.

pan- commented on docs/api/bluetooth/GattClient.md:
```diff
@@ -16,6 +16,14 @@ As a result of the discovery process, the client can start interacting with the
 
+#### ATT_MTU
+
+ATT_MTU is the maximum size of the attribute that can fit in an L2CAP packet. This is independent of the Data Length which controls the over-the-air packet payload size (which is dealt with in the GAP). An L2CAP packet will be fragmented over many such packets if required.
```
I think we can improve the first sentence. What do you think of: 
`The GATT Maximum Transmission Unit (MTU) is the maximum size of a GATT packet exchanged between a client and a server.` Then mention L2CAP as a protocol that allows segmentation and reassembly, ...

pan- commented on docs/api/bluetooth/GattClient.md:
```diff
@@ -16,6 +16,14 @@ As a result of the discovery process, the client can start interacting with the
+
+ATT_MTU is the maximum size of the attribute that can fit in an L2CAP packet. This is independent of the Data Length which controls the over-the-air packet payload size (which is dealt with in the GAP). An L2CAP packet will be fragmented over many such packets if required.
+
+Only `GattClient` can trigger the exchange of ATT_MTU between client and server. The host may trigger the exchange which may also be requested manually using `negotiateAttMtu`. If an exchange happens the biggest value possible across both devices will be used. Negotiation is only a best effort process and does not guarantee a higher value being set. 
```
It would be interesting to detail the negotiation behaviour and clarify the term host.

pan- commented on docs/api/bluetooth/GattServer.md:
```diff
@@ -14,6 +14,14 @@ You must access values of the characteristic and the characteristic descriptor p
 
 You can query the server by invoking the function `areUpdatesEnabled()` to find out if a client has subscribed to a given characteristic's value update.
 
+#### ATT_MTU
```
Same as above

paul-szczepanek-arm committed:
review edits

paul-szczepanek-arm commented:
thanks for the comments, hopefully I have integrated them to your satisfaction

paul-szczepanek-arm committed:
Update GattClient.md

paul-szczepanek-arm committed:
Update GattServer.md

paul-szczepanek-arm committed:
Update GattClient.md

paul-szczepanek-arm committed:
Update GattServer.md

pan- reviewed (approved):
Thanks for the change. It looks good.

Commit:
Edit and rearrange Gap.md

Rearrange content for flow, and edit mostly for precise language and international spelling.

paul-szczepanek-arm commented on docs/api/bluetooth/Gap.md:
```diff
@@ -52,13 +52,15 @@ You may set preferred PHYs (separately for RX and TX) using `setPreferredPhys()`
+#### Data length (over-the-air MTU)
 
-In addition to modulation schemes, MTU (Maximum Transmission Unit) size also strongly affects throughput. Newer controllers will allow you to negotiate bigger MTUs. Since each packet contains overhead bigger packets maximise throughput.
+In addition to modulation schemes, Maximum Transmission Unit (MTU) size also strongly affects throughput. Newer controllers allow you to negotiate bigger MTUs. Because each packet contains overhead, bigger packets maximize throughput.
```
Why is Because superior to Since in this case? Maybe I can try to choose the right one in the future.

AnotherButler replied:
Great question. "Since" shows time passed (for example, "We've been open since 1952"). "Because" shows cause and effect.

pan- replied:
Not trying to open a debate here (you know better than us @AnotherButler) but there is other usages of _since_ (see https://dictionary.cambridge.org/us/grammar/british-grammar/as-because-or-since). Should we avoid these forms in our documentation ?

AnotherButler replied:
Please do. We try to be consistent in our documentation, and [our internal style guide](https://docs.internal.mbed.com/docs/master/style/usage-and-spelling-of-words-you-usually-get-wrong.html) states "Because implies causality; since implies the passing of time. Examples: The program ran because I did it right. Arm has been in business since the 1990s." and "But also note that using "as" when you mean "because" can confuse non-native speakers, so it's best to avoid it." Please don't hesitate to look up words in our style guide if you ever have questions. We can add to it if anything's missing. 

We're also happy to answer questions like this one :)

AnotherButler commented on docs/api/bluetooth/Gap.md:
```diff
@@ -52,6 +52,16 @@ You may set preferred PHYs (separately for RX and TX) using `setPreferredPhys()`
+
+There are two separate MTUs to consider: the `ATT_MTU` (maximum attribute size) and data length. `GattServer` and `GattClient` affect `ATT_MTU`. `Gap` only deals with data length, which is the maximum size of the packet that carries attributes that are fragmented across many such packets.
+
+The default value of data length supported by all controllers is 23 octets. If both controllers support data length extension and a higher value is negotiated, `onDataLengthChange` in the `Gap::EventHandler` will be called.
```
What calls `onDataLengthChange`? You, the user? Or something else?

paul-szczepanek-arm replied:
```suggestion
The default value of data length supported by all controllers is 23 octets. If both controllers support data length extension and a higher value is negotiated, the BLE stack will call `onDataLengthChange` in the `Gap::EventHandler` registered by the user.
```

AnotherButler replied:
👍

paul-szczepanek-arm commented on docs/api/bluetooth/Gap.md:
```diff
@@ -52,6 +52,16 @@ You may set preferred PHYs (separately for RX and TX) using `setPreferredPhys()`
+
+There are two separate MTUs to consider: the `ATT_MTU` (maximum attribute size) and data length. `GattServer` and `GattClient` affect `ATT_MTU`. `Gap` only deals with data length, which is the maximum size of the packet that carries attributes that are fragmented across many such packets.
+
+The default value of data length supported by all controllers is 23 octets. If both controllers support data length extension and a higher value is negotiated, `onDataLengthChange` in the `Gap::EventHandler` will be called.
```
```suggestion
The default value of data length supported by all controllers is 23 octets. If both controllers support data length extension and a higher value is negotiated, the BLE stack will call `onDataLengthChange` in the `Gap::EventHandler` registered by the user.
```

Commit:
Edit GattClient.md

Make minor edits, mostly for active voice.

Commit:
Edit GattServer.md

Edit file, mostly for active voice.

AnotherButler commented on docs/api/bluetooth/GattServer.md:
```diff
@@ -14,6 +16,16 @@ You must access values of the characteristic and the characteristic descriptor p
 
 You can query the server by invoking the function `areUpdatesEnabled()` to find out if a client has subscribed to a given characteristic's value update.
 
+#### Attribute Protocol Maximum Transmission Unit (ATT_MTU)
```
Because this section is an exact duplicate, can we instead replace it with a link to the relevant section in GattClient.md?

paul-szczepanek-arm replied:
OK. The client version is the fuller one as this one shortens the paragraph about what the client does. How do I link?

AnotherButler replied:
The link to GattClient is "../apis/gattclient.html"

paul-szczepanek-arm committed:
Update docs/api/bluetooth/Gap.md

paul-szczepanek-arm committed:
add link to gattclient

paul-szczepanek-arm commented:
Code this documents has been merged

AnotherButler commented:
To clarify, this should only go into 5.12, right? Not 5.11?

paul-szczepanek-arm commented:
yes, PR merged inot master for 5.12

Commit:
Add sentence to GattServer.md

Add one sentence of context in link to GattClient.md.

AnotherButler reviewed (approved):
Thanks for the great work on this 👍
